File: backend/app/schemas/website_content.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field, field_validator

WebsiteSectionKey = Literal["site_identity", "header", "hero", "sections", "footer"]
WebsiteStatus = Literal["draft", "published"]


class WebsiteContentBase(BaseModel):
    section_key: WebsiteSectionKey
    draft_content: dict[str, Any] = Field(default_factory=dict)
    published_content: dict[str, Any] = Field(default_factory=dict)
    status: WebsiteStatus = "draft"
    version: int = 1
    updated_by: Optional[str] = None
    updated_at: Optional[datetime] = None
    published_by: Optional[str] = None
    published_at: Optional[datetime] = None


class WebsiteContentItem(WebsiteContentBase):
    id: str


class WebsiteContentPublicItem(BaseModel):
    section_key: WebsiteSectionKey
    content: dict[str, Any] = Field(default_factory=dict)
    version: int = 1
    published_at: Optional[datetime] = None


class WebsiteContentPublicResponse(BaseModel):
    items: list[WebsiteContentPublicItem]
    generated_at: datetime


class WebsiteContentAdminResponse(BaseModel):
    items: list[WebsiteContentItem]
    generated_at: datetime


class WebsiteContentDraftUpdateRequest(BaseModel):
    draft_content: dict[str, Any] = Field(default_factory=dict)


class WebsiteContentPublishRequest(BaseModel):
    reason: str = Field(..., min_length=1, max_length=500)


class WebsiteContentRollbackRequest(BaseModel):
    version: int = Field(..., ge=1)
    reason: str = Field(..., min_length=1, max_length=500)


class WebsiteContentReorderItem(BaseModel):
    section_key: str = Field(..., min_length=1, max_length=80)
    order: int = Field(..., ge=0, le=100)
    enabled: bool = True


class WebsiteContentReorderRequest(BaseModel):
    items: list[WebsiteContentReorderItem] = Field(default_factory=list)


class WebsiteContentVersionItem(BaseModel):
    id: str
    section_key: WebsiteSectionKey
    version: int
    content: dict[str, Any] = Field(default_factory=dict)
    source: Literal["draft", "published", "rollback"] = "draft"
    created_by: Optional[str] = None
    created_at: datetime
    reason: Optional[str] = None


class WebsiteContentVersionResponse(BaseModel):
    items: list[WebsiteContentVersionItem]
    total: int
    generated_at: datetime

File: frontend/src/types/websiteContent.ts
export type WebsiteSectionKey = 'site_identity' | 'header' | 'hero' | 'sections' | 'footer';

export interface WebsiteMenuItem {
  label: string;
  href: string;
  order: number;
  visible: boolean;
}

export interface SiteIdentityContent {
  site_name: string;
  logo_text: string;
  logo_url?: string;
  favicon_url?: string;
  slogan: string;
}

export interface HeaderContent {
  menu: WebsiteMenuItem[];
  login_label: string;
  primary_cta_label: string;
  authenticated_cta_label: string;
}

export interface HeroContent {
  title: string;
  highlight: string;
  description: string;
  primary_cta_label: string;
  secondary_cta_label: string;
  sticker_image_url?: string;
  upload_enabled: boolean;
  chips: string[];
}

export interface LandingSectionItem {
  key: string;
  title: string;
  eyebrow: string;
  description: string;
  enabled: boolean;
  order: number;
}

export interface BenefitItem {
  title: string;
  description: string;
}

export interface SectionsContent {
  items: LandingSectionItem[];
  benefits: BenefitItem[];
}

export interface FooterLink {
  label: string;
  href: string;
  visible: boolean;
}

export interface FooterContent {
  contact_label: string;
  email: string;
  socials: FooterLink[];
  policies: FooterLink[];
  copyright: string;
}

export interface WebsiteContentBundle {
  site_identity: SiteIdentityContent;
  header: HeaderContent;
  hero: HeroContent;
  sections: SectionsContent;
  footer: FooterContent;
}

export interface WebsiteContentPublicItem {
  section_key: WebsiteSectionKey;
  content: Record<string, unknown>;
  version: number;
  published_at: string | null;
}

export interface WebsiteContentPublicResponse {
  items: WebsiteContentPublicItem[];
  generated_at: string;
}

export interface WebsiteContentAdminItem {
  id: string;
  section_key: WebsiteSectionKey;
  draft_content: Record<string, unknown>;
  published_content: Record<string, unknown>;
  status: 'draft' | 'published';
  version: number;
  updated_by: string | null;
  updated_at: string | null;
  published_by: string | null;
  published_at: string | null;
}

export interface WebsiteContentVersionItem {
  id: string;
  section_key: WebsiteSectionKey;
  version: number;
  content: Record<string, unknown>;
  source: 'draft' | 'published' | 'rollback';
  created_by: string | null;
  created_at: string;
  reason: string | null;
}

export interface WebsiteContentVersionResponse {
  items: WebsiteContentVersionItem[];
  total: number;
  generated_at: string;
}
